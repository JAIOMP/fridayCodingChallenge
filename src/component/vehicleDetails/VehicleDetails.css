.row {
    display: flex;
    align-items: flex-start;
    justify-content: space-between;
    padding: 10px 5px;
}
.row:nth-child(odd) {
    background-color: #ede9e9;
}

.vehicle-details-container {
    display: grid;
    gap: 20px;
    margin: 0 20px;
}

.vehicle-details {
    padding: 10px;
    border: 2px solid lightgrey;
}

.row-label {
    font-weight: 600;
}

@media only screen and (min-width: 720px) {
    .vehicle-details-container {
        grid-template-columns: 1fr 1fr;
    }
}

@media only screen and (min-width: 1024px) {
    .vehicle-details-container {
        grid-template-columns: 1fr 1fr 1fr 1fr;
    }
}
