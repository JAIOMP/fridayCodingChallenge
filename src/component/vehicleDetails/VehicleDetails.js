import React, {useContext, useEffect, useState} from 'react';
import VehicleContext from "../../provider/VehicleDataContext";
import './VehicleDetails.css';

const VehicleDetails = () => {
    const {selectedBrand, selectedModel, error, setError} = useContext(VehicleContext);
    const [vehicleDetails, setVehicleDetails] = useState([]);

    const Row = ({label, value}) => {
        return (<div className="row">
            <span className="row-label">{label}</span>
            <span className="row-value">{value}</span>
        </div>);
    };

    useEffect(() => {
        setError('');
        setVehicleDetails([]);
        fetch(`http://localhost:8080/api/vehicles?make=${selectedBrand}&model=${selectedModel}`)
            .then(res => res.json())
            .then(data => {
                if (data.length === 0) {
                    setError('EMPTY_VEHICLE_DATA');
                    return false;
                }
                setVehicleDetails(data)
            }).catch(() => {
            setError('API_ERROR');
        });
    }, [selectedModel, selectedBrand]);

    return (
        <div className="vehicle-details-container">
            {error === '' && vehicleDetails.map((vehicle, index) => <div className="vehicle-details" key={`${vehicle.make}${index}`}>
                <Row value={vehicle.make} label="Brand"/>
                <Row value={vehicle.model} label="Model"/>
                <Row value={vehicle.enginePowerPS} label="Engine power in ps"/>
                <Row value={vehicle.enginePowerKW} label="Engine power in kw"/>
                <Row value={vehicle.fuelType} label="Fuel type"/>
                <Row value={vehicle.bodyType} label="Body type"/>
                <Row value={vehicle.engineCapacity} label="Engine capacity"/>
            </div>)}
        </div>
    );
};

export default VehicleDetails;
