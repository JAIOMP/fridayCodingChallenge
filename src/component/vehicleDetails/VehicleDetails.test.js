import React from 'react';
import {render, screen, waitFor} from '@testing-library/react';
import VehicleDetails from "./VehicleDetails";
import VehicleContext from "../../provider/VehicleDataContext";

describe('Vehicle Details', () => {
    const mockResponse = [{
        bodyType: "Limousine",
        engineCapacity: 1299,
        enginePowerKW: 44,
        enginePowerPS: 60,
        fuelType: "Benzin",
        make: "FORD",
        model: "Edge",
    }];
    const contextData = {
        setError: jest.fn(),
        selectedBrand: "Ford",
        selectedModel: "Edge",
        error: ""
    };
    beforeEach(() => {
        jest.spyOn(global, 'fetch').mockResolvedValue({
            json: jest.fn().mockResolvedValue(mockResponse)
        })
    });

    it('should render Vehicles', async () => {
        const {container} = render(
            <VehicleContext.Provider value={contextData}>
                <VehicleDetails/>
            </VehicleContext.Provider>
        );

        await screen.findByText("Limousine");

        expect(container).toMatchSnapshot();
    });

    it('should not render vehicle details if error is present', async () => {
        jest.spyOn(global, 'fetch').mockImplementation(() => Promise.reject("API is down"));
        render(
            <VehicleContext.Provider value={contextData}>
                <VehicleDetails/>
            </VehicleContext.Provider>
        );
        await waitFor(() => {
            expect(contextData.setError).toHaveBeenCalledTimes(2);
            expect(contextData.setError).toHaveBeenNthCalledWith(1, "");
            expect(contextData.setError).toHaveBeenNthCalledWith(2, "API_ERROR");
        })
    });
});
