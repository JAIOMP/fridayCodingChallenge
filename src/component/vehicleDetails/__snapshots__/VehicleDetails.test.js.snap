// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vehicle Details should render Vehicles 1`] = `
<div>
  <div
    class="vehicle-details-container"
  >
    <div
      class="vehicle-details"
    >
      <div
        class="row"
      >
        <span
          class="row-label"
        >
          Brand
        </span>
        <span
          class="row-value"
        >
          FORD
        </span>
      </div>
      <div
        class="row"
      >
        <span
          class="row-label"
        >
          Model
        </span>
        <span
          class="row-value"
        >
          Edge
        </span>
      </div>
      <div
        class="row"
      >
        <span
          class="row-label"
        >
          Engine power in ps
        </span>
        <span
          class="row-value"
        >
          60
        </span>
      </div>
      <div
        class="row"
      >
        <span
          class="row-label"
        >
          Engine power in kw
        </span>
        <span
          class="row-value"
        >
          44
        </span>
      </div>
      <div
        class="row"
      >
        <span
          class="row-label"
        >
          Fuel type
        </span>
        <span
          class="row-value"
        >
          Benzin
        </span>
      </div>
      <div
        class="row"
      >
        <span
          class="row-label"
        >
          Body type
        </span>
        <span
          class="row-value"
        >
          Limousine
        </span>
      </div>
      <div
        class="row"
      >
        <span
          class="row-label"
        >
          Engine capacity
        </span>
        <span
          class="row-value"
        >
          1299
        </span>
      </div>
    </div>
  </div>
</div>
`;
